src/scanner.ts

```diff
@@ -12,35 +12,48 @@ export interface ScanResult {
   ruleName: string;
   description: string;
   total: number;
-  ids: string[];
+  resources: {id: string}[];
 }
 
 export class Scanner {
-  private _rules: Rule[] = [];
-
-  async scan(ruleType: Rule['type'], target: string) {
-    if (!this._rules.length) await this.loadRulesFromFile();
-    return this._executeRules(ruleType, target);
+  async scan(context: RuleContext, target: RuleContext['target']) {
+    switch (context.type) {
+      case 'resourceGraph': {
+        context.target = target as ResourceGraphRuleContext['target'];
+        return ResourceGraphRule.execute(context);
+      }
+      case 'dummy': {
+        context.target = target as DummyRuleContext['target'];
+        return DummyRule.execute(context);
+      }
+    }
   }
 
-  async loadRulesFromFile(filePath = '../../rules.json') {
-    const absPath = path.join(__dirname, filePath);
-    const data = await fsPromises.readFile(absPath, 'utf8');
-    this._rules = JSON.parse(data);
+  getContextByType(type: RuleContext['type'], contexts: RuleContext[]) {
+    return contexts.filter(r => r.type === type)[0];
```

Comment:
Is it true that this array will only contain one instance/block of rules for a given type? Seems like we can't enforce this, so we just silently drop all the 2nd+ rules on the ground

We should enforce via JSON structure (ex: property name == rule type), combine arrays of the same type, throw warnings/errors if multiple sections of the same type are encountered, etc.